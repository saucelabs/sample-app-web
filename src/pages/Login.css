.login_logo {
  text-align: center;
  padding: 30px 0;
  background: url("../assets/img/SwagLabs_logo.png") no-repeat center center;
  margin-top: 30px;
}
.login_wrapper {
  padding: 30px 0;
  margin: 0px auto;
}
.login_wrapper-inner {
  overflow: hidden;
}
.login_wrapper-inner,
.login_credentials_wrap-inner {
  width: 70%;
  max-width: 720px;
  padding: 15px 30px 30px;
  margin: 0 auto;
  border: 2px solid #ededed;
}
.form_column {
  float: none;
  width: 100%;
}
.bot_column {
  float: none;
  width: 100%;
  height: 100%;
  background: url("../assets/img/Login_Bot_graphic.png") no-repeat center center;
  min-height: 260px;
  min-width: 300px;
}
.login-button {
  margin: 20px auto;
  background: #fff;
  border: 2px solid #ce2a2b;
  font-size: 18px;
  width: 100%;
  max-width: 400px;
  padding: 10px;
  text-align: center;
  color: #ce2a2b;
  outline: none;
}
.login-button:hover {
  background: #efefef;
  cursor: pointer;
}
.login-button:active {
  background: #ccc;
}
.login_credentials {
  clear: both;
  line-height: 24px;
}
.login_credentials {
  float: left;
  width: 48%;
}
.login_password {
  float: right;
  width: 48%;
}
.login_credentials_wrap-inner {
  background-color: #efefef;
  overflow: hidden;
}
@media only screen and (min-width: 900px) {
  .form_column {
    float: left;
    width: 48%;
    padding-top: 30px;
  }
  .bot_column {
    float: right;
    width: 48%;
    background: url("../assets/img/Login_Bot_graphic.png") no-repeat center
      center;
    min-height: 260px;
    min-width: 300px;
  }
}
