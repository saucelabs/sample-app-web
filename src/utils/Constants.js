export const FONT_FAMILY = "Roboto, Arial, Helvetica, sans-serif";
export const VALID_USERNAMES = [
  "standard_user",
  "locked_out_user",
  "problem_user",
  "performance_glitch_user",
  "error_user",
  "visual_user",
];
export const VALID_PASSWORD = "secret_sauce";
export const ROUTES = {
  LOGIN: "/",
  INVENTORY: "/inventory.html",
  INVENTORY_LIST: "/inventory-item.html",
  CART: "/cart.html",
  CHECKOUT_STEP_ONE: "/checkout-step-one.html",
  CHECKOUT_STEP_TWO: "/checkout-step-two.html",
  CHECKOUT_COMPLETE: "/checkout-complete.html",
};
export const CART_CONTENTS = "cart-contents";
export const SESSION_USERNAME = "session-username";
