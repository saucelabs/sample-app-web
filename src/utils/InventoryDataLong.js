export const InventoryDataLong = [
  {
    id: 0,
    name: "Sauce Labs Bike Light",
    desc: `A red light isn't the desired state in testing but it sure helps when riding your bike at night. Water-resistant with 3 lighting modes, 1 AAA battery included.`,
    price: 9.99,
    image_url: "bike-light-1200x1500.jpg",
  },
  {
    id: 1,
    name: "Sauce Labs Bolt T-Shirt (XS)",
    desc: `Get your testing superhero on with the Sauce Labs bolt T-shirt. From American Apparel, 100% ringspun combed cotton, heather gray with red bolt.`,
    price: 15.99,
    image_url: "bolt-shirt-1200x1500.jpg",
  },
  {
    id: 2,
    name: "Sauce Labs Onesie",
    desc: `Rib snap infant onesie for the junior automation engineer in development. Reinforced 3-snap bottom closure, two-needle hemmed sleeved and bottom won't unravel.`,
    price: 7.99,
    image_url: "red-onesie-1200x1500.jpg",
  },
  {
    id: 3,
    name: "Test.allTheThings() T-Shirt (Red) (XS)",
    desc: `This classic Sauce Labs t-shirt is perfect to wear when cozying up to your keyboard to automate a few tests. Super-soft and comfy ringspun combed cotton.`,
    price: 15.99,
    image_url: "red-tatt-1200x1500.jpg",
  },
  {
    id: 4,
    name: "Sauce Labs Backpack",
    desc: `carry.allTheThings() with the sleek, streamlined Sly Pack that melds uncompromising style with unequaled laptop and tablet protection.`,
    price: 29.99,
    image_url: "sauce-backpack-1200x1500.jpg",
  },
  {
    id: 5,
    name: "Sauce Labs Fleece Jacket (XS)",
    desc: `It's not every day that you come across a midweight quarter-zip fleece jacket capable of handling everything from a relaxing day outdoors to a busy day at the office.`,
    price: 49.99,
    image_url: "sauce-pullover-1200x1500.jpg",
  },
  {
    id: 6,
    name: "Sauce Labs Fleece Jacket (S)",
    desc: `It's not every day that you come across a midweight quarter-zip fleece jacket capable of handling everything from a relaxing day outdoors to a busy day at the office.`,
    price: 49.99,
    image_url: "sauce-pullover-1200x1500.jpg",
  },
  {
    id: 7,
    name: "Sauce Labs Fleece Jacket (M)",
    desc: `It's not every day that you come across a midweight quarter-zip fleece jacket capable of handling everything from a relaxing day outdoors to a busy day at the office.`,
    price: 49.99,
    image_url: "sauce-pullover-1200x1500.jpg",
  },
  {
    id: 8,
    name: "Sauce Labs Fleece Jacket (L)",
    desc: `It's not every day that you come across a midweight quarter-zip fleece jacket capable of handling everything from a relaxing day outdoors to a busy day at the office.`,
    price: 49.99,
    image_url: "sauce-pullover-1200x1500.jpg",
  },
  {
    id: 9,
    name: "Sauce Labs Fleece Jacket (XL)",
    desc: `It's not every day that you come across a midweight quarter-zip fleece jacket capable of handling everything from a relaxing day outdoors to a busy day at the office.`,
    price: 49.99,
    image_url: "sauce-pullover-1200x1500.jpg",
  },
  {
    id: 10,
    name: "Sauce Labs Fleece Jacket (XXL)",
    desc: `It's not every day that you come across a midweight quarter-zip fleece jacket capable of handling everything from a relaxing day outdoors to a busy day at the office.`,
    price: 49.99,
    image_url: "sauce-pullover-1200x1500.jpg",
  },
  {
    id: 11,
    name: "Test.allTheThings() T-Shirt (Red) (S)",
    desc: `This classic Sauce Labs t-shirt is perfect to wear when cozying up to your keyboard to automate a few tests. Super-soft and comfy ringspun combed cotton.`,
    price: 15.99,
    image_url: "red-tatt-1200x1500.jpg",
  },

  {
    id: 12,
    name: "Test.allTheThings() T-Shirt (Red) (M)",
    desc: `This classic Sauce Labs t-shirt is perfect to wear when cozying up to your keyboard to automate a few tests. Super-soft and comfy ringspun combed cotton.`,
    price: 15.99,
    image_url: "red-tatt-1200x1500.jpg",
  },

  {
    id: 13,
    name: "Test.allTheThings() T-Shirt (Red) (L)",
    desc: `This classic Sauce Labs t-shirt is perfect to wear when cozying up to your keyboard to automate a few tests. Super-soft and comfy ringspun combed cotton.`,
    price: 15.99,
    image_url: "red-tatt-1200x1500.jpg",
  },

  {
    id: 14,
    name: "Test.allTheThings() T-Shirt (Red) (L)",
    desc: `This classic Sauce Labs t-shirt is perfect to wear when cozying up to your keyboard to automate a few tests. Super-soft and comfy ringspun combed cotton.`,
    price: 15.99,
    image_url: "red-tatt-1200x1500.jpg",
  },

  {
    id: 15,
    name: "Test.allTheThings() T-Shirt (Red) (XL)",
    desc: `This classic Sauce Labs t-shirt is perfect to wear when cozying up to your keyboard to automate a few tests. Super-soft and comfy ringspun combed cotton.`,
    price: 15.99,
    image_url: "red-tatt-1200x1500.jpg",
  },

  {
    id: 16,
    name: "Test.allTheThings() T-Shirt (Red) (XXL)",
    desc: `This classic Sauce Labs t-shirt is perfect to wear when cozying up to your keyboard to automate a few tests. Super-soft and comfy ringspun combed cotton.`,
    price: 15.99,
    image_url: "red-tatt-1200x1500.jpg",
  },
  {
    id: 17,
    name: "Sauce Labs Bolt T-Shirt (M)",
    desc: `Get your testing superhero on with the Sauce Labs bolt T-shirt. From American Apparel, 100% ringspun combed cotton, heather gray with red bolt.`,
    price: 15.99,
    image_url: "bolt-shirt-1200x1500.jpg",
  },
  {
    id: 18,
    name: "Sauce Labs Bolt T-Shirt (L)",
    desc: `Get your testing superhero on with the Sauce Labs bolt T-shirt. From American Apparel, 100% ringspun combed cotton, heather gray with red bolt.`,
    price: 15.99,
    image_url: "bolt-shirt-1200x1500.jpg",
  },
  {
    id: 19,
    name: "Sauce Labs Bolt T-Shirt (XL)",
    desc: `Get your testing superhero on with the Sauce Labs bolt T-shirt. From American Apparel, 100% ringspun combed cotton, heather gray with red bolt.`,
    price: 15.99,
    image_url: "bolt-shirt-1200x1500.jpg",
  },
  {
    id: 20,
    name: "Sauce Labs Bolt T-Shirt (XXL)",
    desc: `Get your testing superhero on with the Sauce Labs bolt T-shirt. From American Apparel, 100% ringspun combed cotton, heather gray with red bolt.`,
    price: 15.99,
    image_url: "bolt-shirt-1200x1500.jpg",
  },
  {
    id: 21,
    name: "Sauce Labs Bolt T-Shirt (S)",
    desc: `Get your testing superhero on with the Sauce Labs bolt T-shirt. From American Apparel, 100% ringspun combed cotton, heather gray with red bolt.`,
    price: 15.99,
    image_url: "bolt-shirt-1200x1500.jpg",
  },
];
