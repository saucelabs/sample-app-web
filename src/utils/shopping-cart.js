export class ShoppingCart {
  static addItem(itemId) {
    // pull out our current cart contents
    const curContents = ShoppingCart.getCartContents();

    /* istanbul ignore else */
    if (curContents.indexOf(itemId) < 0) {
      // Item's not yet present - add it now
      curContents.push(itemId);

      // We modified our cart, so store it now
      ShoppingCart.setCartContents(curContents);
    }
  }

  static removeItem(itemId) {
    // pull out our current cart contents
    const curContents = ShoppingCart.getCartContents();
    const itemIndex = curContents.indexOf(itemId);

    /* istanbul ignore else */
    if (itemIndex >= 0) {
      // Remove this item from the array
      curContents.splice(itemIndex, 1);

      // We modified our cart, so store it now
      ShoppingCart.setCartContents(curContents);
    }
  }

  static isItemInCart(itemId) {
    // pull out our current cart contents
    const curContents = ShoppingCart.getCartContents();

    // If the item is in the array, return true
    return curContents.indexOf(itemId) >= 0;
  }

  static getCartContents() {
    // pull out our current cart contents
    let curContents = window.localStorage.getItem("cart-contents");

    // Make an empty list if this is the first item
    if (curContents == null) {
      curContents = [];
    } else {
      // We have an existing cart, so deserialize it now since localStorage stores in JSON strings
      curContents = JSON.parse(curContents);
    }

    return curContents;
  }

  static setCartContents(newContents) {
    window.localStorage.setItem("cart-contents", JSON.stringify(newContents));

    // Notify our listeners
    /* istanbul ignore next */
    ShoppingCart.LISTENERS.forEach((curListener) => {
      curListener.forceUpdate();
    });
  }

  static resetCart() {
    window.localStorage.removeItem("cart-contents");

    // Notify our listeners
    /* istanbul ignore next */
    ShoppingCart.LISTENERS.forEach((curListener) => {
      curListener.forceUpdate();
    });
  }

  static registerCartListener(handler) {
    ShoppingCart.LISTENERS.push(handler);
  }
}

ShoppingCart.LISTENERS = [];
