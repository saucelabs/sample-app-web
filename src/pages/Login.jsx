import React, { useEffect } from "react";
import { withRouter } from "react-router-dom";
import { useState } from "react";
import "./Login.css";
import {
  isLockedOutUser,
  setCredentials,
  verifyCredentials,
} from "../utils/Credentials";
import { ROUTES } from "../utils/Constants";
import InputError, { INPUT_TYPES } from "../components/InputError";
import SubmitButton from "../components/SubmitButton";
import ErrorMessage from "../components/ErrorMessage";

function Login(props) {
  const { history, location } = props;
  const [error, setError] = useState("");
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");

  useEffect(() => {
    if (location.state) {
      return setError(
        `You can only access '${location.state.from.pathname}' when you are logged in.`
      );
    }
  }, [location.state]);

  const dismissError = () => {
    setError("");
  };

  const handleSubmit = (evt) => {
    evt.preventDefault();
    if (!username) {
      return setError("Username is required");
    }

    if (!password) {
      return setError("Password is required");
    }

    if (verifyCredentials(username, password)) {
      // If we're here, we have a username and password.
      // Store the username in our cookies.
      setCredentials(username, password);
      // Catch our locked-out user and bail out
      if (isLockedOutUser()) {
        return setError("Sorry, this user has been locked out.");
      }

      // Redirect!
      history.push(ROUTES.INVENTORY);
    } else {
      return setError(
        "Username and password do not match any user in this service"
      );
    }

    return "";
  };

  const handleUserChange = (evt) => {
    setUsername(evt.target.value);
  };

  const handlePassChange = (evt) => {
    setPassword(evt.target.value);
  };

  return (
    <div>
      <div className="login_logo" />

      <div className="login_wrapper">
        <div className="login_wrapper-inner">
          <div id="login_button_container" className="form_column">
            <div className="login-box">
              <form onSubmit={handleSubmit}>
                <InputError
                  isError={Boolean(error)}
                  type={INPUT_TYPES.TEXT}
                  value={username}
                  onChange={handleUserChange}
                  testId="username"
                  placeholder="Username"
                  // Custom
                  id="user-name"
                  name="user-name"
                  autoCorrect="off"
                  autoCapitalize="none"
                />
                <InputError
                  isError={Boolean(error)}
                  type={INPUT_TYPES.PASSWORD}
                  value={password}
                  onChange={handlePassChange}
                  testId="password"
                  placeholder="Password"
                  // Custom
                  autoCorrect="off"
                  autoCapitalize="none"
                />
                <ErrorMessage
                  isError={Boolean(error)}
                  errorMessage={`Epic sadface: ${error}`}
                  onClick={dismissError}
                />
                <SubmitButton
                  // `btn_action` has no style function
                  // but is there for backwards compatibility
                  customClass="btn_action"
                  testId="login-button"
                  value="Login"
                />
              </form>
            </div>
          </div>

          <div className="bot_column" />
        </div>
        <div className="login_credentials_wrap">
          <div className="login_credentials_wrap-inner">
            <div id="login_credentials" className="login_credentials">
              <h4>Accepted usernames are:</h4>
              standard_user
              <br />
              locked_out_user
              <br />
              problem_user
              <br />
              performance_glitch_user
              <br />
            </div>
            <div className="login_password">
              <h4>Password for all users:</h4>
              secret_sauce
            </div>
          </div>
        </div>
      </div>
    </div>
  );
}

export default withRouter(Login);
