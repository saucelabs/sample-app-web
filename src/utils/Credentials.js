import Cookies from "js-cookie";
import { SESSION_USERNAME, VALID_PASSWORD, VALID_USERNAMES } from "./Constants";

/**
 * Verify the credentials
 *
 * @param {string} username
 * @param {string} password
 *
 * @return {boolean}
 */
export function verifyCredentials(username, password) {
  if (password !== VALID_PASSWORD) {
    return false;
  }

  return VALID_USERNAMES.includes(username);
}

/**
 * Store the data in our cookies
 *
 * @param {string} username
 *
 * @param {string} password
 */
export function setCredentials(username, password) {
  let date = new Date();
  date.setTime(date.getTime() + 10 * 60 * 1000);

  Cookies.set(SESSION_USERNAME, username, { expires: date });
}

/**
 * Remove the credentials
 */
export function removeCredentials() {
  Cookies.remove(SESSION_USERNAME);
}

/**
 * Return current logged username
 *
 * @return {string | undefined}
 */
export function currentUser() {
  return Cookies.get(SESSION_USERNAME);
}

/**
 * Check if this is a problem user
 *
 * @return {boolean}
 */
export function isProblemUser() {
  return Cookies.get(SESSION_USERNAME) === "problem_user";
}

/**
 * Check if this is a performance user
 *
 * @return {boolean}
 */
export function isPerformanceGlitchUser() {
  return Cookies.get(SESSION_USERNAME) === "performance_glitch_user";
}

/**
 * Check if this a logged out user
 *
 * @return {boolean}
 */
export function isLockedOutUser() {
  return Cookies.get(SESSION_USERNAME) === "locked_out_user";
}

/**
 * Check if this is an error user
 *
 * @return {boolean}
 */
export function isErrorUser() {
  return Cookies.get(SESSION_USERNAME) === "error_user";
}

/**
 * Check if the user is logged in with a valid username
 *
 * @return {boolean}
 */
export function isLoggedIn() {
  const sessionUsername = Cookies.get(SESSION_USERNAME);
  const isValidUsername = VALID_USERNAMES.includes(sessionUsername);

  return isValidUsername && sessionUsername !== "locked_out_user";
}

/**
 * Check if this is a visual user
 *
 * @return {boolean}
 */
export function isVisualUser() {
  return Cookies.get(SESSION_USERNAME) === "visual_user";
}
