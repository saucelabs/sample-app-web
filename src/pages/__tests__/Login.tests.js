import React, { useState as useStateMock } from "react";
import { shallow } from "enzyme";
import Login from "../Login";
import * as Credentials from "../../utils/Credentials";

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: jest.fn(),
  useEffect: (f) => f(),
}));

describe("Login", () => {
  const setState = jest.fn();

  beforeEach(() => {
    useStateMock.mockImplementation((init) => [init, setState]);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should render correctly", () => {
    const wrapper = shallow(<Login.WrappedComponent history location />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should be able to dismiss the error", () => {
    const wrapper = shallow(<Login.WrappedComponent history location />);
    const ErrorMessage = wrapper.find("ErrorMessage");

    ErrorMessage.at(0).simulate("click");

    expect(setState).toHaveBeenCalledWith("");
  });

  it("should update the username state when data is typed through the handleUserChange", () => {
    const wrapper = shallow(<Login.WrappedComponent history location />);
    const InputErrors = wrapper.find("InputError");
    const value = "username";

    InputErrors.at(0).simulate("change", { target: { value } });

    expect(setState).toHaveBeenCalledWith(value);
  });

  it("should update the password state when data is typed through the handlePassChange", () => {
    const wrapper = shallow(<Login.WrappedComponent history location />);
    const InputErrors = wrapper.find("InputError");
    const value = "password";

    InputErrors.at(1).simulate("change", { target: { value } });

    expect(setState).toHaveBeenCalledWith(value);
  });

  it("should set the no username state when the login button is being pressed", () => {
    const wrapper = shallow(<Login.WrappedComponent history location />);
    const form = wrapper.find("form").at(0);
    form.simulate("submit", {
      preventDefault() {},
    });

    expect(setState).toHaveBeenCalledWith("Username is required");
  });

  it("should set the no password state when the login button is being pressed", () => {
    useStateMock
      // error useState
      .mockImplementationOnce((stateValue) => [
        (stateValue = "error"),
        setState,
      ])
      // username useState
      .mockImplementationOnce((stateValue) => [
        (stateValue = "username"),
        setState,
      ]);

    const wrapper = shallow(<Login.WrappedComponent history location />);
    const InputErrors = wrapper.find("InputError");
    InputErrors.at(0).simulate("change", { target: { value: "username" } });
    const form = wrapper.find("form").at(0);
    form.simulate("submit", {
      preventDefault() {},
    });

    expect(setState).toHaveBeenCalledTimes(2);
    expect(setState).toHaveBeenNthCalledWith(2, "Password is required");
  });

  it("should set the locked out state when the locked out credentials are used", () => {
    useStateMock
      // error useState
      .mockImplementationOnce((stateValue) => [
        (stateValue = "error"),
        setState,
      ])
      // username useState
      .mockImplementationOnce((stateValue) => [
        (stateValue = "username"),
        setState,
      ])
      // password useState
      .mockImplementationOnce((stateValue) => [
        (stateValue = "password"),
        setState,
      ]);

    const verifyCredentialsSpy = jest.spyOn(Credentials, "verifyCredentials");
    const isLockedOutUserSpy = jest.spyOn(Credentials, "isLockedOutUser");
    verifyCredentialsSpy.mockReturnValueOnce(true);
    isLockedOutUserSpy.mockReturnValueOnce(true);

    const wrapper = shallow(<Login.WrappedComponent history location />);
    const InputErrors = wrapper.find("InputError");
    InputErrors.at(0).simulate("change", { target: { value: "username" } });
    InputErrors.at(1).simulate("change", { target: { value: "password" } });
    const form = wrapper.find("form").at(0);
    form.simulate("submit", {
      preventDefault() {},
    });

    expect(verifyCredentialsSpy).toHaveBeenCalledTimes(1);
    expect(isLockedOutUserSpy).toHaveBeenCalledTimes(1);
    expect(setState).toHaveBeenCalledTimes(3);
    expect(setState).toHaveBeenNthCalledWith(
      3,
      "Sorry, this user has been locked out."
    );

    verifyCredentialsSpy.mockClear();
    isLockedOutUserSpy.mockClear();
  });

  it("should set the data does not match state when incorrect credentials are used", () => {
    useStateMock
      // error useState
      .mockImplementationOnce((stateValue) => [
        (stateValue = "error"),
        setState,
      ])
      // username useState
      .mockImplementationOnce((stateValue) => [
        (stateValue = "username"),
        setState,
      ])
      // password useState
      .mockImplementationOnce((stateValue) => [
        (stateValue = "password"),
        setState,
      ]);

    const verifyCredentialsSpy = jest.spyOn(Credentials, "verifyCredentials");
    verifyCredentialsSpy.mockReturnValueOnce(false);

    const wrapper = shallow(<Login.WrappedComponent history location />);
    const InputErrors = wrapper.find("InputError");
    InputErrors.at(0).simulate("change", { target: { value: "username" } });
    InputErrors.at(1).simulate("change", { target: { value: "password" } });
    const form = wrapper.find("form").at(0);
    form.simulate("submit", {
      preventDefault() {},
    });

    expect(verifyCredentialsSpy).toHaveBeenCalledTimes(1);
    expect(setState).toHaveBeenCalledTimes(3);
    expect(setState).toHaveBeenNthCalledWith(
      3,
      "Username and password do not match any user in this service"
    );

    verifyCredentialsSpy.mockClear();
  });

  it("should redirect when valid credentials are used", () => {
    useStateMock
      // error useState
      .mockImplementationOnce((stateValue) => [
        (stateValue = "error"),
        setState,
      ])
      // username useState
      .mockImplementationOnce((stateValue) => [
        (stateValue = "username"),
        setState,
      ])
      // password useState
      .mockImplementationOnce((stateValue) => [
        (stateValue = "password"),
        setState,
      ]);

    const verifyCredentialsSpy = jest.spyOn(Credentials, "verifyCredentials");
    const isLockedOutUserSpy = jest.spyOn(Credentials, "isLockedOutUser");
    verifyCredentialsSpy.mockReturnValueOnce(true);
    isLockedOutUserSpy.mockReturnValueOnce(false);

    const historyProps = { push: jest.fn() };
    const wrapper = shallow(
      <Login.WrappedComponent history={historyProps} location />
    );
    const InputErrors = wrapper.find("InputError");
    InputErrors.at(0).simulate("change", { target: { value: "username" } });
    InputErrors.at(1).simulate("change", { target: { value: "password" } });
    const form = wrapper.find("form").at(0);
    form.simulate("submit", {
      preventDefault() {},
    });

    expect(verifyCredentialsSpy).toHaveBeenCalledTimes(1);
    expect(isLockedOutUserSpy).toHaveBeenCalledTimes(1);
    expect(setState).toHaveBeenCalledTimes(2);
    expect(historyProps.push).toBeCalledWith("/inventory.html");

    verifyCredentialsSpy.mockClear();
    isLockedOutUserSpy.mockClear();
  });

  it("should set the redirect error state", () => {
    const value = "redirected-path";
    const locationProps = { state: { from: { pathname: value } } };
    shallow(<Login.WrappedComponent history location={locationProps} />);

    expect(setState).toHaveBeenCalledWith(
      `You can only access '${value}' when you are logged in.`
    );
  });
});
