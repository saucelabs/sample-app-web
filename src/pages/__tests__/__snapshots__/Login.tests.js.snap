// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render correctly 1`] = `
<div>
  <div
    className="login_logo"
  />
  <div
    className="login_wrapper"
  >
    <div
      className="login_wrapper-inner"
    >
      <div
        className="form_column"
        id="login_button_container"
      >
        <div
          className="login-box"
        >
          <form
            onSubmit={[Function]}
          >
            <InputError
              autoCapitalize="none"
              autoCorrect="off"
              id="user-name"
              isError={false}
              name="user-name"
              onChange={[Function]}
              placeholder="Username"
              testId="username"
              type="text"
              value=""
            />
            <InputError
              autoCapitalize="none"
              autoCorrect="off"
              isError={false}
              onChange={[Function]}
              placeholder="Password"
              testId="password"
              type="password"
              value=""
            />
            <ErrorMessage
              errorMessage="Epic sadface: "
              isError={false}
              onClick={[Function]}
            />
            <SubmitButton
              customClass="btn_action"
              testId="login-button"
              value="Login"
            />
          </form>
        </div>
      </div>
      <div
        className="bot_column"
      />
    </div>
    <div
      className="login_credentials_wrap"
    >
      <div
        className="login_credentials_wrap-inner"
      >
        <div
          className="login_credentials"
          id="login_credentials"
        >
          <h4>
            Accepted usernames are:
          </h4>
          standard_user
          <br />
          locked_out_user
          <br />
          problem_user
          <br />
          performance_glitch_user
          <br />
        </div>
        <div
          className="login_password"
        >
          <h4>
            Password for all users:
          </h4>
          secret_sauce
        </div>
      </div>
    </div>
  </div>
</div>
`;
